ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/MiniProtocol/ObjectDiffusion/Inbound/V2/Decision.hs

```diff
@@ -0,0 +1,345 @@
+      -- If we do this change, we could add an assert in `V2.hs` that whenever pdNumIdsToReq == 0, then pdNumIdsToAck == 0 as well
+      -- /!\ We should also revise documentation in V2.md accordingly
+
+      pdCanPipelineIdsRequests = not . StrictSeq.null $ dpsOutstandingFifo'
+
+      peerDecision =
+        PeerDecision
+          { pdNumIdsToAck
+          , pdNumIdsToReq
+          , pdCanPipelineIdsRequests
+          , pdObjectsToReqIds = Set.empty -- we don't decide this here
+          , pdStatus = DecisionUnread
+          }
+     in
+      ( Map.insert peerAddr peerDecision decisionsAcc
+      , Map.insert peerAddr (strictSeqToSet idsToAck) peerToIdsToAck
+      )
+
+-- | Order peers randomly based on the provided RNG.
+-- We do that to avoid biasing the decision logic towards the peers that
+-- would happen to be first in the map default ordering, as that could
+-- be exploited by an adversary.
+orderPeers ::
+  StdGen ->
+  Map peerAddr (DecisionPeerState objectId object) ->
+  [(peerAddr, DecisionPeerState objectId object)]
+orderPeers rng peerStates =
+  let peerPairs = Map.toList peerStates
+   in shuffle' peerPairs (length peerPairs) rng
+
+data DownloadPickState peerAddr objectId
+  = DownloadPickState
+  { totalNumObjectsToReq :: !NumObjectsReq
+  , objectMultiplicity :: ObjectMultiplicity
+  , peersToObjectsToReq :: Map peerAddr (Set objectId)
+  }
+
+-- | This function could just be pure if it hadn't be for the rng used to order peers
+pickObjectsToReq ::
+  forall peerAddr objectId object.
+  ( Ord peerAddr
+  , Ord objectId
+  ) =>
+  StdGen ->
+  (objectId -> Bool) ->
+  DecisionPolicy ->
+  DecisionGlobalState peerAddr objectId object ->
+  -- | Frozen peers and their previous decisions
+  Map peerAddr (PeerDecision objectId object) ->
+  -- | map from peer to the set of ids that will be acked for that peer on next requestIds
+  -- we should treat these ids as not available anymore for the purpose of picking objects to request
+  Map peerAddr (Set objectId) ->
+  -- | new global state (with just RNG updated), and objects to request from each peer
+  Map peerAddr (Set objectId)
+pickObjectsToReq
+  rng
+  poolHasObject
+  DecisionPolicy
+    { dpMaxNumObjectsInflightPerPeer
+    , dpMaxNumObjectsInflightTotal
+    , dpTargetObjectRedundancy
+    }
+  DecisionGlobalState
+    { dgsPeerStates
+    }
+  frozenPeersToDecisions
+  peerToIdsToAck =
+    peersToObjectsToReq
+   where
+    -- We order the peers that are not currently executing a decision
+    orderedPeers = orderPeers rng (dgsPeerStates `Map.withoutKeys` Map.keysSet frozenPeersToDecisions)
+
+    -- We want to map each objectId to the sorted list of peers that can provide it
+    -- For each peer we also indicate how many objects it has in flight at the moment
+    -- We filter out here the objects that are already in pool
+    objectsToSortedProviders :: Map objectId [(peerAddr, NumObjectsReq)]
```

Comment:
Is the order of the elements important here? I think you could use a set instead of a list here (`peerAddr` already implements `Ord`), and turn the (O(n*2)) fold below (L217-L221) into O(n.log(n)).

Reply:
Yes, we need to preserve order, that's why I'm using a list :)